File: admin/manage_produk.py
from flask import Blueprint, request, jsonify
from model import query_fetchall, query_fetchone, query_execute
from functools import wraps
import jwt
import os
from werkzeug.utils import secure_filename
from config import config

produk_bp = Blueprint('produk', __name__)


# Reuse decorator dari dashboard - harusnya dibuat jadi middleware terpisah
# tapi buat sekarang duplikat dulu biar ga ribet import
def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get('Authorization', '')
        if not auth_header.startswith('Bearer '):
            return jsonify({'success': False, 'message': 'Login dulu ya'}), 401
        token = auth_header.split(' ')[1]
        try:
            payload = jwt.decode(token, config.JWT_SECRET, algorithms=['HS256'])
            request.admin = payload
        except:
            return jsonify({'success': False, 'message': 'Token tidak valid'}), 401
        return f(*args, **kwargs)
    return decorated


def allowed_file(filename):
    """Cek apakah ekstensi file diizinkan"""
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in config.ALLOWED_EXTENSIONS


@produk_bp.route('/', methods=['GET'])
@login_required
def get_all_produk():
    """
    Ambil semua produk dengan filter dan pagination
    Query params: page, limit, kategori, search
    """
    page = int(request.args.get('page', 1))
    limit = int(request.args.get('limit', 3))  # 3 per halaman sesuai desain
    kategori_id = request.args.get('kategori', None)
    search = request.args.get('search', '').strip()
    
    offset = (page - 1) * limit
    
    # Bangun query dinamis berdasarkan filter
    where_clauses = []
    params = []
    
    if kategori_id:
        where_clauses.append("p.kategori_id = %s")
        params.append(kategori_id)
    
    if search:
        where_clauses.append("(p.nama_produk LIKE %s OR p.kode_produk LIKE %s)")
        params.extend([f'%{search}%', f'%{search}%'])
    
    where_sql = ""
    if where_clauses:
        where_sql = "WHERE " + " AND ".join(where_clauses)
    
    # Hitung total data dulu (buat pagination)
    count_sql = f"""
        SELECT COUNT(*) as total 
        FROM produk p 
        LEFT JOIN kategori k ON p.kategori_id = k.id
        {where_sql}
    """
    total_data = query_fetchone(count_sql, params)
    total = total_data['total'] if total_data else 0
    
    # Ambil data dengan pagination
    data_sql = f"""
        SELECT 
            p.id, p.kode_produk, p.nama_produk, 
            k.nama_kategori as kategori,
            p.harga, p.stok, p.foto, p.status,
            p.created_at
        FROM produk p
        LEFT JOIN kategori k ON p.kategori_id = k.id
        {where_sql}
        ORDER BY p.created_at DESC
        LIMIT %s OFFSET %s
    """
    params.extend([limit, offset])
    produk_list = query_fetchall(data_sql, params)
    
    # Format harga biar lebih enak dibaca
    for p in produk_list:
        p['harga_format'] = f"Rp {int(p['harga']):,}".replace(',', '.')
    
    return jsonify({
        'success': True,
        'data': produk_list,
        'pagination': {
            'total': total,
            'page': page,
            'limit': limit,
            'total_halaman': (total + limit - 1) // limit
        }
    }), 200


@produk_bp.route('/kategori', methods=['GET'])
@login_required
def get_kategori():
    """Ambil semua kategori buat dropdown filter"""
    kategori = query_fetchall("SELECT * FROM kategori ORDER BY nama_kategori")
    return jsonify({'success': True, 'data': kategori}), 200


@produk_bp.route('/<int:id>', methods=['GET'])
@login_required
def get_produk_by_id(id):
    """Ambil detail satu produk berdasarkan ID"""
    produk = query_fetchone("""
        SELECT p.*, k.nama_kategori as kategori
        FROM produk p
        LEFT JOIN kategori k ON p.kategori_id = k.id
        WHERE p.id = %s
    """, (id,))
    
    if not produk:
        return jsonify({'success': False, 'message': 'Produk tidak ditemukan'}), 404
    
    return jsonify({'success': True, 'data': produk}), 200


@produk_bp.route('/', methods=['POST'])
@login_required
def tambah_produk():
    """
    Tambah produk baru
    Form data: kode_produk, nama_produk, kategori_id, harga, stok, foto (file)
    """
    # Ambil data dari form (bukan JSON karena ada upload file)
    kode = request.form.get('kode_produk', '').strip()
    nama = request.form.get('nama_produk', '').strip()
    kategori_id = request.form.get('kategori_id')
    harga = request.form.get('harga')
    stok = request.form.get('stok', 0)
    
    # Validasi field wajib
    if not kode or not nama or not kategori_id or not harga:
        return jsonify({'success': False, 'message': 'Semua field wajib diisi'}), 400
    
    # Cek kode produk udah ada belum
    existing = query_fetchone("SELECT id FROM produk WHERE kode_produk = %s", (kode,))
    if existing:
        return jsonify({'success': False, 'message': f'Kode produk {kode} sudah digunakan'}), 400
    
    # Handle upload foto
    foto_filename = None
    if 'foto' in request.files:
        foto = request.files['foto']
        if foto.filename != '' and allowed_file(foto.filename):
            foto_filename = secure_filename(f"{kode}_{foto.filename}")
            foto.save(os.path.join(config.UPLOAD_FOLDER, foto_filename))
    
    # Tentukan status berdasarkan stok
    status = 'habis' if int(stok) == 0 else 'tersedia'
    
    result = query_execute("""
        INSERT INTO produk (kode_produk, nama_produk, kategori_id, harga, stok, foto, status)
        VALUES (%s, %s, %s, %s, %s, %s, %s)
    """, (kode, nama, kategori_id, harga, stok, foto_filename, status))
    
    if result:
        return jsonify({'success': True, 'message': 'Produk berhasil ditambahkan', 'id': result}), 201
    else:
        return jsonify({'success': False, 'message': 'Gagal menambahkan produk'}), 500


@produk_bp.route('/<int:id>', methods=['PUT'])
@login_required
def update_produk(id):
    """Update data produk yang sudah ada"""
    
    # Cek produk ada dulu
    existing = query_fetchone("SELECT * FROM produk WHERE id = %s", (id,))
    if not existing:
        return jsonify({'success': False, 'message': 'Produk tidak ditemukan'}), 404
    
    nama = request.form.get('nama_produk', existing['nama_produk']).strip()
    kategori_id = request.form.get('kategori_id', existing['kategori_id'])
    harga = request.form.get('harga', existing['harga'])
    stok = request.form.get('stok', existing['stok'])
    
    # Handle foto baru kalau ada
    foto_filename = existing['foto']
    if 'foto' in request.files:
        foto = request.files['foto']
        if foto.filename != '' and allowed_file(foto.filename):
            # Hapus foto lama kalau ada
            if existing['foto']:
                old_path = os.path.join(config.UPLOAD_FOLDER, existing['foto'])
                if os.path.exists(old_path):
                    os.remove(old_path)
            
            foto_filename = secure_filename(f"{existing['kode_produk']}_{foto.filename}")
            foto.save(os.path.join(config.UPLOAD_FOLDER, foto_filename))
    
    status = 'habis' if int(stok) == 0 else 'tersedia'
    
    query_execute("""
        UPDATE produk 
        SET nama_produk=%s, kategori_id=%s, harga=%s, stok=%s, foto=%s, status=%s
        WHERE id=%s
    """, (nama, kategori_id, harga, stok, foto_filename, status, id))
    
    return jsonify({'success': True, 'message': 'Produk berhasil diupdate'}), 200


@produk_bp.route('/<int:id>', methods=['DELETE'])
@login_required
def hapus_produk(id):
    """Hapus produk berdasarkan ID"""
    
    existing = query_fetchone("SELECT * FROM produk WHERE id = %s", (id,))
    if not existing:
        return jsonify({'success': False, 'message': 'Produk tidak ditemukan'}), 404
    
    # Hapus foto dari storage juga
    if existing['foto']:
        foto_path = os.path.join(config.UPLOAD_FOLDER, existing['foto'])
        if os.path.exists(foto_path):
            os.remove(foto_path)
    
    query_execute("DELETE FROM produk WHERE id = %s", (id,))
    
    return jsonify({'success': True, 'message': 'Produk berhasil dihapus'}), 200

File: admin/manage_produk.js
/**
 * manage-produk.js
 * Handle CRUD produk di halaman Manage Produk
 */

// Guard - cek login
if (!getToken()) {
    window.location.href = '/admin/login.html';
}

// State halaman
let currentPage = 1;
let totalHalaman = 1;
let currentEditId = null;
let currentHapusId = null;
let searchTimeout = null;

// Referensi elemen DOM
const produkBody = document.getElementById('produkBody');
const totalBadge = document.getElementById('totalBadge');
const paginationInfo = document.getElementById('paginationInfo');
const paginationBtns = document.getElementById('paginationBtns');
const searchInput = document.getElementById('searchInput');
const kategoriFilter = document.getElementById('kategoriFilter');
const modalOverlay = document.getElementById('modalOverlay');
const modalHapusOverlay = document.getElementById('modalHapusOverlay');
const modalTitle = document.getElementById('modalTitle');
const btnSimpan = document.getElementById('btnSimpan');
const toast = document.getElementById('toast');

// ---- Fungsi Utilitas ----

function tampilkanToast(pesan, tipe = 'sukses') {
    toast.textContent = pesan;
    toast.className = `toast ${tipe} show`;
    setTimeout(() => toast.classList.remove('show'), 3000);
}

function bukaModal() {
    modalOverlay.classList.add('show');
}

function tutupModal() {
    modalOverlay.classList.remove('show');
    resetForm();
}

function bukaModalHapus(id, nama) {
    currentHapusId = id;
    document.getElementById('hapusNama').textContent = nama;
    modalHapusOverlay.classList.add('show');
}

function tutupModalHapus() {
    modalHapusOverlay.classList.remove('show');
    currentHapusId = null;
}

function resetForm() {
    document.getElementById('inputKode').value = '';
    document.getElementById('inputNama').value = '';
    document.getElementById('inputKategori').value = '';
    document.getElementById('inputHarga').value = '';
    document.getElementById('inputStok').value = '0';
    document.getElementById('inputFoto').value = '';
    currentEditId = null;
    modalTitle.textContent = 'Tambah Produk';
    btnSimpan.textContent = 'Simpan Produk';
    document.getElementById('inputKode').disabled = false;
}

function getKategoriClass(kategori) {
    const k = kategori?.toLowerCase().replace(/[^a-z]/g, '');
    if (k === 'coffee') return 'coffee';
    if (k === 'noncoffee') return 'non-coffee';
    if (k === 'pastry') return 'pastry';
    return '';
}

function getFotoUrl(fotoFilename) {
    if (!fotoFilename) {
        return 'https://images.unsplash.com/photo-1495474472287-4d71bcdd2085?w=60&h=60&fit=crop';
    }
    return `/admin/uploads/${fotoFilename}`;
}

// ---- Load Data ----

async function loadKategori() {
    const result = await api.get('/produk/kategori');
    if (!result || !result.data.success) return;

    const options = result.data.data.map(k =>
        `<option value="${k.id}">${k.nama_kategori}</option>`
    ).join('');

    // Isi dropdown filter
    kategoriFilter.innerHTML = '<option value="">Semua Kategori</option>' + options;

    // Isi dropdown di modal
    document.getElementById('inputKategori').innerHTML =
        '<option value="">Pilih kategori...</option>' + options;
}

async function loadProduk() {
    produkBody.innerHTML = `<tr><td colspan="8" class="loading-text">Memuat data...</td></tr>`;

    const search = searchInput.value.trim();
    const kategori = kategoriFilter.value;

    let endpoint = `/produk/?page=${currentPage}&limit=3`;
    if (search) endpoint += `&search=${encodeURIComponent(search)}`;
    if (kategori) endpoint += `&kategori=${kategori}`;

    const result = await api.get(endpoint);

    if (!result || !result.data.success) {
        produkBody.innerHTML = `<tr><td colspan="8" class="loading-text">Gagal memuat produk</td></tr>`;
        return;
    }

    const { data, pagination } = result.data;
    totalHalaman = pagination.total_halaman;

    // Update info total
    totalBadge.textContent = `Total: ${pagination.total} Produk`;
    paginationInfo.textContent = `Menampilkan ${(currentPage - 1) * 3 + 1}-${Math.min(currentPage * 3, pagination.total)} dari ${pagination.total} produk`;

    if (data.length === 0) {
        produkBody.innerHTML = `<tr><td colspan="8" class="loading-text">Tidak ada produk yang ditemukan</td></tr>`;
        renderPaginasi(0, 0);
        return;
    }

    produkBody.innerHTML = data.map(p => `
        <tr>
            <td>
                <img 
                    class="foto-produk"
                    src="${getFotoUrl(p.foto)}"
                    alt="${p.nama_produk}"
                    onerror="this.src='https://images.unsplash.com/photo-1495474472287-4d71bcdd2085?w=60&h=60&fit=crop'"
                >
            </td>
            <td><span class="kode-badge">${p.kode_produk}</span></td>
            <td><strong>${p.nama_produk}</strong></td>
            <td>
                <span class="kategori-badge ${getKategoriClass(p.kategori)}">
                    ${p.kategori || '-'}
                </span>
            </td>
            <td>${p.harga_format || 'Rp ' + Number(p.harga).toLocaleString('id-ID')}</td>
            <td class="${p.stok <= 3 ? 'stok-rendah' : ''}">${p.stok}</td>
            <td>
                <span class="status-badge ${p.status}">
                    ${p.status === 'tersedia' ? 'Tersedia' : 'Habis'}
                </span>
            </td>
            <td>
                <div class="aksi-btns">
                    <button class="btn-edit" onclick="editProduk(${p.id})" title="Edit">
                        <svg width="14" height="14" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2">
                            <path d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"/>
                            <path d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"/>
                        </svg>
                    </button>
                    <button class="btn-hapus" onclick="konfirmasiHapus(${p.id}, '${p.nama_produk}')" title="Hapus">
                        <svg width="14" height="14" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2">
                            <polyline points="3 6 5 6 21 6"/>
                            <path d="M19 6l-1 14H6L5 6M10 11v6M14 11v6"/>
                            <path d="M9 6V4h6v2"/>
                        </svg>
                    </button>
                </div>
            </td>
        </tr>
    `).join('');

    renderPaginasi(pagination.total, pagination.total_halaman);
}

function renderPaginasi(total, totalPage) {
    if (total === 0) {
        paginationBtns.innerHTML = '';
        return;
    }

    let html = `
        <button class="page-btn" onclick="gantiHalaman(${currentPage - 1})" ${currentPage === 1 ? 'disabled' : ''}>‹</button>
    `;

    for (let i = 1; i <= totalPage; i++) {
        // Tampilkan halaman 1, terakhir, dan sekitar halaman aktif
        if (i === 1 || i === totalPage || Math.abs(i - currentPage) <= 1) {
            html += `<button class="page-btn ${i === currentPage ? 'active' : ''}" onclick="gantiHalaman(${i})">${i}</button>`;
        } else if (Math.abs(i - currentPage) === 2) {
            html += `<span class="page-btn" style="cursor:default">...</span>`;
        }
    }

    html += `
        <button class="page-btn" onclick="gantiHalaman(${currentPage + 1})" ${currentPage === totalPage ? 'disabled' : ''}>›</button>
    `;

    paginationBtns.innerHTML = html;
}

function gantiHalaman(hal) {
    if (hal < 1 || hal > totalHalaman) return;
    currentPage = hal;
    loadProduk();
}

// ---- CRUD ----

// Klik tombol edit - isi form dengan data produk yang dipilih
async function editProduk(id) {
    const result = await api.get(`/produk/${id}`);
    if (!result || !result.data.success) {
        tampilkanToast('Gagal memuat data produk', 'error');
        return;
    }

    const p = result.data.data;
    currentEditId = id;

    document.getElementById('inputKode').value = p.kode_produk;
    document.getElementById('inputKode').disabled = true; // kode gak boleh diubah
    document.getElementById('inputNama').value = p.nama_produk;
    document.getElementById('inputKategori').value = p.kategori_id;
    document.getElementById('inputHarga').value = p.harga;
    document.getElementById('inputStok').value = p.stok;

    modalTitle.textContent = 'Edit Produk';
    btnSimpan.textContent = 'Update Produk';
    bukaModal();
}

function konfirmasiHapus(id, nama) {
    bukaModalHapus(id, nama);
}

// Simpan produk (tambah atau update)
async function simpanProduk() {
    const kode = document.getElementById('inputKode').value.trim();
    const nama = document.getElementById('inputNama').value.trim();
    const kategori = document.getElementById('inputKategori').value;
    const harga = document.getElementById('inputHarga').value;
    const stok = document.getElementById('inputStok').value;
    const fotoFile = document.getElementById('inputFoto').files[0];

    // Validasi sederhana
    if (!nama || !kategori || !harga) {
        tampilkanToast('Lengkapi semua field yang wajib diisi', 'error');
        return;
    }

    // Pakai FormData karena ada file upload
    const formData = new FormData();
    if (!currentEditId) formData.append('kode_produk', kode);
    formData.append('nama_produk', nama);
    formData.append('kategori_id', kategori);
    formData.append('harga', harga);
    formData.append('stok', stok);
    if (fotoFile) formData.append('foto', fotoFile);

    btnSimpan.disabled = true;
    btnSimpan.textContent = 'Menyimpan...';

    let result;
    if (currentEditId) {
        result = await api.putForm(`/produk/${currentEditId}`, formData);
    } else {
        result = await api.postForm('/produk/', formData);
    }

    btnSimpan.disabled = false;
    btnSimpan.textContent = currentEditId ? 'Update Produk' : 'Simpan Produk';

    if (result && result.data.success) {
        tampilkanToast(result.data.message, 'sukses');
        tutupModal();
        loadProduk();
    } else {
        tampilkanToast(result?.data?.message || 'Terjadi kesalahan', 'error');
    }
}

// Hapus produk setelah konfirmasi
async function hapusProduk() {
    if (!currentHapusId) return;

    const result = await api.delete(`/produk/${currentHapusId}`);

    if (result && result.data.success) {
        tampilkanToast('Produk berhasil dihapus', 'sukses');
        tutupModalHapus();

        // Kalo halaman ini jadi kosong, balik ke halaman sebelumnya
        if (currentPage > 1) currentPage--;
        loadProduk();
    } else {
        tampilkanToast(result?.data?.message || 'Gagal menghapus produk', 'error');
    }
}

// ---- Event Listeners ----

document.getElementById('btnTambah').addEventListener('click', () => {
    resetForm();
    bukaModal();
});

document.getElementById('modalClose').addEventListener('click', tutupModal);
document.getElementById('btnBatal').addEventListener('click', tutupModal);
document.getElementById('hapusClose').addEventListener('click', tutupModalHapus);
document.getElementById('hapusBatal').addEventListener('click', tutupModalHapus);
document.getElementById('btnSimpan').addEventListener('click', simpanProduk);
document.getElementById('hapusKonfirm').addEventListener('click', hapusProduk);

// Tutup modal kalo klik di luar area modal
modalOverlay.addEventListener('click', (e) => {
    if (e.target === modalOverlay) tutupModal();
});

modalHapusOverlay.addEventListener('click', (e) => {
    if (e.target === modalHapusOverlay) tutupModalHapus();
});

// Search dengan debounce biar gak request terus setiap ketikan
searchInput.addEventListener('input', () => {
    clearTimeout(searchTimeout);
    searchTimeout = setTimeout(() => {
        currentPage = 1;
        loadProduk();
    }, 400);
});

// Filter kategori
kategoriFilter.addEventListener('change', () => {
    currentPage = 1;
    loadProduk();
});

// Logout
document.getElementById('btnLogout').addEventListener('click', () => {
    if (confirm('Yakin mau logout?')) {
        api.post('/auth/logout', {});
        removeToken();
        window.location.href = '/admin/login.html';
    }
});

// ---- Inisialisasi ----
async function init() {
    await loadKategori();
    await loadProduk();
}

init();